internal/node/package.go

```diff
@@ -2,22 +2,92 @@ package node
 
 import (
 	"encoding/json"
+	"github.com/rs/zerolog/log"
 	"os"
+	"path/filepath"
 )
 
 type Package struct {
-	Dependencies    map[string]string `json:"dependencies,omitempty"`
-	DevDependencies map[string]string `json:"devDependencies,omitempty"`
+	Dependencies     map[string]string `json:"dependencies,omitempty"`
+	DevDependencies  map[string]string `json:"devDependencies,omitempty"`
+	PeerDependencies map[string]string `json:"peerDependencies,omitempty"`
 }
 
 func PackageFromFile(filename string) (Package, error) {
 	var p Package
 
 	fd, err := os.Open(filename)
 	if err != nil {
-		return p, err 
+		return p, err
 	}
 
 	err = json.NewDecoder(fd).Decode(&p)
 	return p, err
 }
+
+// Requirements returns a list of all root dependencies for the given packages, including themselves.
+func Requirements(root string, pkgs ...string) []string {
+	rootDeps := make(map[string]struct{})
+
+	log.Info().Msgf("Getting requirements for npm dependencies %s", pkgs)
+	for _, pkg := range pkgs {
+		requirements(rootDeps, root, pkg, true)
+	}
+
+	var deps []string
+	for k := range rootDeps {
+		deps = append(deps, k)
+	}
+	return deps
+}
+
+// requirements recursively traverses the dependency tree for the given package, adding all root (shared) dependencies
+// to the rootDeps map.
+func requirements(rootDeps map[string]struct{}, root, pkg string, rootDep bool) {
+	if _, ok := rootDeps[pkg]; ok {
+		return
+	}
+
+	log.Debug().Msgf("Getting requirements for %s", pkg)
+	p, err := PackageFromFile(filepath.Join(root, pkg, "package.json"))
+	if err != nil {
+		if os.IsNotExist(err) {
+			// likely an unmet peer dependency, not necessarily an issue
+			log.Debug().Msgf("Unmet requirement %s", pkg)
+			return
+		}
+		log.Err(err).Msgf("Failed to get requirements for %s", pkg)
+		return
+	}
+	if rootDep {
+		rootDeps[pkg] = struct{}{}
+	}
+
+	var requiredDeps []string
+	for k := range p.Dependencies {
+		requiredDeps = append(requiredDeps, k)
+	}
+	for k := range p.PeerDependencies {
+		requiredDeps = append(requiredDeps, k)
+	}
```

Comment:
I'm actually not sure about the inclusion of peer dependencies. There are cases where the inclusion is beneficial or even required. In other cases, well, let's say you have a plugin that depends on cypress. That plugin most likely specifies the cypress as a peer dependency. See the [code coverage plugin](https://github.com/cypress-io/code-coverage/blob/master/package.json). If we inspect the peer dependencies, that'd also mean we'd include cypress itself as part of the upload and then ofc all of its dependencies as well.

My current thought: We may need to include peer dependencies, but do a hard-exclude on framework based dependencies, so that we don't include cypress/playwright/testcafe.

Reply:
Excluding framework packages seems necessary to me. Aside from it possibly being a peer dependency, what about the case that a user explicitly (and erroneously) declares it as a dependency in their configuration?

Reply:
Seems like we need to ignore the core framework packages from pretty much anywhere 🔪